In [25]:
import pandas as pd
import numpy as np

df = pd.read_csv('adult.csv')
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,?,77053,HS-grad,9,Widowed,?,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,?,186061,Some-college,10,Widowed,?,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K


In [26]:
#Task 1
df.isnull().sum()   

df.replace('?', np.nan, inplace=True)

df.isnull().sum()   

age                  0
workclass         1836
fnlwgt               0
education            0
education.num        0
marital.status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital.gain         0
capital.loss         0
hours.per.week       0
native.country     583
income               0
dtype: int64

In [27]:
#Task 2
missing_count = df.isnull().sum()
missing_pct = df.isnull().mean() * 100

missing_summary = pd.DataFrame({'count': missing_count, 'percent': missing_pct})
missing_summary = missing_summary[missing_summary['count'] > 0].sort_values('count', ascending=False)
missing_summary

,count,percent
occupation,1843,5.660146
workclass,1836,5.638647
native.country,583,1.790486


In [7]:
# Task 3
df['workclass'] = df['workclass'].fillna('Unknown')
df['occupation'] = df['occupation'].fillna('Unknown')
df['native.country'] = df['native.country'].fillna(df['native.country'].mode()[0])

In [18]:
df.head()

,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,Unknown,77053,HS-grad,9,Widowed,Unknown,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,Unknown,186061,Some-college,10,Widowed,Unknown,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K


In [10]:
#Task 4
print(df.duplicated().sum())


print(df.duplicated(subset=[c for c in df.columns if c != 'income']).sum())


24
25


In [11]:
#Task 5
for col in ['education', 'marital.status', 'native.country']:
    print(col, df[col].unique())

df['education'] = df['education'].str.strip()
df['marital.status'] = df['marital.status'].str.strip()
df['native.country'] = df['native.country'].str.strip()

education ['HS-grad' 'Some-college' '7th-8th' '10th' 'Doctorate' 'Prof-school'
 'Bachelors' 'Masters' '11th' 'Assoc-acdm' 'Assoc-voc' '1st-4th' '5th-6th'
 '12th' '9th' 'Preschool']
marital.status ['Widowed' 'Divorced' 'Separated' 'Never-married' 'Married-civ-spouse'
 'Married-spouse-absent' 'Married-AF-spouse']
native.country ['United-States' 'Mexico' 'Greece' 'Vietnam' 'China' 'Taiwan' 'India'
 'Philippines' 'Trinadad&Tobago' 'Canada' 'South' 'Holand-Netherlands'
 'Puerto-Rico' 'Poland' 'Iran' 'England' 'Germany' 'Italy' 'Japan' 'Hong'
 'Honduras' 'Cuba' 'Ireland' 'Cambodia' 'Peru' 'Nicaragua'
 'Dominican-Republic' 'Haiti' 'El-Salvador' 'Hungary' 'Columbia'
 'Guatemala' 'Jamaica' 'Ecuador' 'France' 'Yugoslavia' 'Scotland'
 'Portugal' 'Laos' 'Thailand' 'Outlying-US(Guam-USVI-etc)']


In [12]:
print(df['education'])

0             HS-grad
1             HS-grad
2        Some-college
3             7th-8th
4        Some-college
             ...     
32556    Some-college
32557      Assoc-acdm
32558         HS-grad
32559         HS-grad
32560         HS-grad
Name: education, Length: 32561, dtype: object


In [15]:
#Task 6
df.describe()

,age,fnlwgt,education.num,capital.gain,capital.loss,hours.per.week
count,32561.000000,3.256100e+04,32561.000000,32561.000000,32561.000000,32561.000000
mean,38.581647,1.897784e+05,10.080679,1077.648844,87.303830,40.437456
std,13.640433,1.055500e+05,2.572720,7385.292085,402.960219,12.347429
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.370510e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


age: mean = 38.58, median = 37.0, close, only a 1.5 year gap, so age isn't heavily skewed by outliers.

hours-per-week: mean =40.43, median = 40.0,since 40 hrs/week is the dominant standard work schedule for most respondents; the small mean-median gap suggests a mild right tail

In [14]:
df.describe(include='object')

,workclass,education,marital.status,occupation,relationship,race,sex,native.country,income
count,32561,32561,32561,32561,32561,32561,32561,32561,32561
unique,9,16,7,15,6,5,2,41,2
top,Private,HS-grad,Married-civ-spouse,Prof-specialty,Husband,White,Male,United-States,<=50K
freq,22696,10501,14976,4140,13193,27816,21790,29753,24720


In [29]:
#Task 7
df['occupation'].value_counts()


occupation
Prof-specialty       4140
Craft-repair         4099
Exec-managerial      4066
Adm-clerical         3770
Sales                3650
Other-service        3295
Machine-op-inspct    2002
Transport-moving     1597
Handlers-cleaners    1370
Farming-fishing       994
Tech-support          928
Protective-serv       649
Priv-house-serv       149
Armed-Forces            9
Name: count, dtype: int64

In [33]:
df['sex'].value_counts(normalize=True) * 100

sex
Male      66.920549
Female    33.079451
Name: proportion, dtype: float64

In [31]:
df['income'].value_counts(normalize=True) * 100

income
<=50K    75.919044
>50K     24.080956
Name: proportion, dtype: float64

In [34]:
# Task 8
df.groupby('education')['education.num'].unique()

education
10th             [6]
11th             [7]
12th             [8]
1st-4th          [2]
5th-6th          [3]
7th-8th          [4]
9th              [5]
Assoc-acdm      [12]
Assoc-voc       [11]
Bachelors       [13]
Doctorate       [16]
HS-grad          [9]
Masters         [14]
Preschool        [1]
Prof-school     [15]
Some-college    [10]
Name: education.num, dtype: object

In [35]:
#Task 9
import sqlite3

conn = sqlite3.connect('adult_income.db')
df.to_sql('adult_income', conn, index=False, if_exists='replace')

query_df = pd.read_sql_query('SELECT * FROM adult_income WHERE age > 30', conn)

print(query_df.shape)   
print(df.shape)         

(21989, 15)
(32561, 15)


## Data Profiling Summary

**Dataset overview**
32,561 rows and 15 columns from the UCI Adult Income dataset

**Data quality issues found**
- Missing values hidden as literal "?" strings (not real NaN) in three columns:
  occupation , workclass, and native.country 
- 24 exact duplicate rows; 25 rows duplicate on all features except income
- No whitespace inconsistencies found in education, marital.status, or
  native.country in this version of the file.

**Key observations**
- Income classes are imbalanced: 75.9% earn <=50K vs 24.1% earning >50K
- Sex distribution is skewed: 66.9% Male vs 33.1% Female.
- Most common occupation is Prof-specialty (4,140), narrowly ahead of
  Craft-repair (4,099) and Exec-managerial (4,066).
- education and education.num are perfectly consistent, no mismatches.

**Cleaning decisions made**
- Replaced "?" with NaN to expose true missingness.
- Imputed workclass and occupation with "Unknown" rather than dropping,
  to preserve ~5.6% of rows for other features.
- Imputed native.country with its mode (United-States) given its low
  missing rate (1.79%) and heavy skew toward one category.


In [36]:
df.to_csv("adult-cleaned.csv", index= False)
